File: android/app/src/main/java/com/aurea/aurea/ads/UmpConsent.kt
package com.aurea.aurea.ads

import android.app.Activity
import com.google.android.ump.ConsentInformation
import com.google.android.ump.ConsentRequestParameters
import com.google.android.ump.UserMessagingPlatform
import java.util.concurrent.atomic.AtomicBoolean

/**
 * O consentimento oficial do Google (UMP), comum a qualquer provedor: ele grava
 * o consentimento no padrão IAB TCF, que o AdMob, o LevelPlay e as redes
 * mediadas leem. Nada de consentimento presumido: `aoDecidir(true)` só quando
 * `canRequestAds()` permite.
 */
object UmpConsent {
    fun pedir(activity: Activity, aoDecidir: (podePedir: Boolean) -> Unit) {
        val uma = AtomicBoolean(false)
        val decidir = { ok: Boolean -> if (uma.compareAndSet(false, true)) aoDecidir(ok) }
        val consent: ConsentInformation = UserMessagingPlatform.getConsentInformation(activity.applicationContext)
        consent.requestConsentInfoUpdate(activity, ConsentRequestParameters.Builder().build(), {
            // O formulário oficial só aparece se a região/usuário exigir.
            UserMessagingPlatform.loadAndShowConsentFormIfRequired(activity) { _ -> decidir(consent.canRequestAds()) }
        }, { _ ->
            // Sem rede para atualizar: vale o que já estava decidido antes.
            decidir(consent.canRequestAds())
        })
        // Sessões anteriores já decidiram: pode seguir em paralelo (recomendação do Google).
        if (consent.canRequestAds()) decidir(true)
    }
}

File: engine/platform/ios/app/Ads/UmpConsent.swift
// =============================================================================
//  Aurea iOS — o consentimento oficial do Google (UMP), comum a qualquer
//  provedor: grava o consentimento no padrão IAB TCF, que o LevelPlay e as
//  redes mediadas leem. Nada de consentimento presumido: `decide(true)` só
//  quando `canRequestAds` permite. (O mesmo `UmpConsent` do Android.)
// =============================================================================
import UIKit
import UserMessagingPlatform

enum UmpConsent {
    static func request(from vc: UIViewController, decide: @escaping (_ canRequestAds: Bool) -> Void) {
        var answered = false
        let answer: (Bool) -> Void = { ok in
            DispatchQueue.main.async { if !answered { answered = true; decide(ok) } }
        }
        ConsentInformation.shared.requestConsentInfoUpdate(with: RequestParameters()) { error in
            // Sem rede para atualizar: vale o que já estava decidido antes.
            if error != nil { answer(ConsentInformation.shared.canRequestAds); return }
            // O formulário oficial só aparece se a região/usuário exigir.
            DispatchQueue.main.async {
                ConsentForm.loadAndPresentIfRequired(from: vc) { _ in answer(ConsentInformation.shared.canRequestAds) }
            }
        }
        // Sessões anteriores já decidiram: pode seguir em paralelo (recomendação do Google).
        if ConsentInformation.shared.canRequestAds { answer(true) }
    }
}
